### PR TITLE
Mascara toggle

### PR DESCRIPTION
**Pull request by frankiebee**
this will allow developers to switch between the mascara provider and the extension provider 
so that you dont have to disable the extension
simply type `web3.currentProvider.mascara = true` and the providers will switch and use the mascara provider

### CODE REVIEW

File: app/scripts/lib/mascara-detection-usage.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  const head = document.getElementsByTagName('head')[0]
+  const headContent = head.children
+  const childrenLength = headContent.length
+  // the default is to serve the extensions provider
+  self.mascara = false
+  self.extensionProvider = provider
+  for (let index = 0; index < childrenLength; index++) {
+    const src = headContent[index].getAttribute('src')
+    if (src === 'MASCARA_PROXY_ORIGIN') {
```
Comment (kumavis) [inline app/scripts/lib/mascara-detection-usage.js]: lets make this a CONSTANT_VARIABLE and then set it at the top or somewhere

we could also prolly do this via process.env.ENV_VAR

Comment (kumavis) [inline app/scripts/lib/mascara-detection-usage.js]: or add attribute to the src tag like `metamascara=true`

Comment (kumavis) [inline app/scripts/lib/mascara-detection-usage.js]: or maybe better to check `window.mascara` or something

File: gulpfile.js [hunk 2]
```diff
@@ -311,6 +313,7 @@ function bundleTask(opts) {
       .pipe(source(opts.filename))
       // inject variables into bundle
       .pipe(replace('\'GULP_METAMASK_DEBUG\'', debug))
+      .pipe(replace('\'MASCARA_PROXY_ORIGIN\'', MASCARA_PROXY_ORIGIN))
```
Comment (kumavis) [inline gulpfile.js]: should be able to inject via envify or something (this feels a bit hacky, tho i know theres precedent the line above)

**Review by kumavis — CHANGES_REQUESTED** [review]
better to use `mascara` or `metamask` global var for detecting instead of looking for the script

File: app/scripts/lib/mascara-detection-usage.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+module.exports = function MascaraDetectionUsageSetup (extensionProvider) {
+  let mascara = false
+  if (global.mascaraProvider) {
+    let networkVersion
+    global.mascaraProvider.publicConfigStore.subscribe((state) => {
+      if (!global.web3.currentProvider.mascara) return
```
Comment (kumavis) [inline app/scripts/lib/mascara-detection-usage.js]: please rename to `isMascaraActive`

File: app/scripts/inpage.js [hunk 4]
```diff
@@ -21,23 +22,22 @@ var metamaskStream = new LocalMessageDuplexStream({
...
 // export global web3, with usage-detection
 setupDappAutoReload(web3, inpageProvider.publicConfigStore)
-
+console.log('MetaMask - injected web3')
 // set web3 defaultAccount
 
 inpageProvider.publicConfigStore.subscribe(function (state) {
-  web3.eth.defaultAccount = state.selectedAddress
+  // if not using the mascara provider
+  if (!inpageProvider.mascara) web3.eth.defaultAccount = state.selectedAddress
```
Comment (kumavis) [inline app/scripts/inpage.js]: please rename to `isMascaraActive`

File: mascara/src/mascara.js [hunk 5]
```diff
@@ -11,6 +11,9 @@ console.log('MASCARA_ORIGIN:', MASCARA_ORIGIN)
 const provider = setupProvider({
   mascaraUrl: MASCARA_ORIGIN + '/proxy/',
 })
+provider.isMascara = true
```
Comment (kumavis) [inline mascara/src/mascara.js]: remove

File: gulpfile.js [hunk 6]
```diff
@@ -123,6 +123,7 @@ gulp.task('manifest:production', function() {
   ],{base: './dist/'})
   .pipe(gulpif(!debug,jsoneditor(function(json) {
     json.background.scripts = ["scripts/background.js"]
+    json["content_scripts"][1]["exclude_matches"] = [`${MASCARA_ORIGIN}*`]
```
Comment (kumavis) [inline gulpfile.js]: we should skip injection on prod mascara even when we're in dev mode

**Review by kumavis — CHANGES_REQUESTED** [review]

File: gulpfile.js (hunk 6 above)
Comment (frankiebee) [inline gulpfile.js]: content_scripts is an array with an object already in it. The object already in it is the one we want to write too

**Review by kumavis — APPROVED** [review]

File: app/scripts/lib/mascara-proxy-provider.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
+module.exports = function setupMascaraProxyProvider (extensionProvider) {
+  let isMascaraActive = false
+  if (global.mascaraProvider) {
```
Comment (kumavis) [inline app/scripts/lib/mascara-proxy-provider.js]: we should just abort if `global.mascaraProvide` is falsy, yes?

Comment (frankiebee) [inline app/scripts/lib/mascara-proxy-provider.js]: sure i will change it but... food for thought: what if the race condition shows up it wont be obvious till you send a tx that mascara isnt global. Left this way it throws an error if you expected it to be their and it's not.

**Comment by tmashuang** [comment]
- [ ] clicking multiple times opens multiple transactions pop-ups
- [ ] Metamask pop up shows account screen after rejecting the transaction.
       STR: click Async Tx -> reject -> click anywhere on ZeroClient

**Comment by Zanibas** [comment]
Why are we closing this?

**Pull request closed without merging**